text/3543-patchable-function-entry.md

```diff
@@ -0,0 +1,267 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`patchable-function-entry` provides configurable nop padding before function symbols and after function symbols but before any generated code. We refer to the former as `prefix` padding and the latter as `entry` padding. For example, if we had a function `f` with `prefix` set to 3 and `entry` to 2, we'd expect to see:
+
+```
+nop
+nop
+nop
+f:
+nop
+nop
+// Code goes here
+```
+
+To set this for all functions in a crate, use `-C patchable-function-entry=nop_count=m,offset=n` where `nop_count = prefix + entry`, and `offset = prefix`. Usually, you'll want to copy this value from a corresponding `-fpatchable-function-entry=` being passed to the C compiler in your project.
+
+To set this for a specific function, use `#[patchable(prefix = m, entry = n)]` to pad with m nops before the symbol and n after the symbol, but before the prelude. This will override the flag value. To disable padding for a specific function, for example because it is part of the instrumentation framework, use `#[unpatchable]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`patchable-function-entry` provides configurable nop padding before function symbols and after function symbols but before any generated code. We refer to the former as `prefix` padding and the latter as `entry` padding. For example, if we had a function `f` with `prefix` set to 3 and `entry` to 2, we'd expect to see:
+
+```
+f_pad:
+nop
+nop
+nop
+f:
+nop
+nop
+// Code goes here
+```
+
+Nop padding may not be supported on all architectures. As of the time of writing, support includes:
+
+- aarch64
+- aarch64\_be
+- loongarch32
+- loongarch64
+- riscv32
+- riscv64
+- i686
+- x86\_64
+
+`f_pad` addresses for every padded symbol are aggregated in the `__patchable_function_entries` section of the resulting object.
+This is not a real symbol, just a collected location.
+
+## Compiler flag `-C patchable-function-entry`
+
+This flag comes in two forms:
+
+- `-C patchable-function-entry=nop_count=m,offset=n`
+- `-C patchable-function-entry=nop_count=m`
+
+In the latter, offset is assumed to be zero. `nop_count` must be greater than or equal to `offset`, or it will be rejected.
+
+If unspecified, the current behavior is maintained, which is equivalent to `nop_count=0` here.
+
+This flag sets the default nop padding for all functions in the crate. Notably, this default *only applies to codegenned functions*. If a function is monomorphized during the compilation of another crate or any similar scenario, it will use the default from that crate's compilation. In most cases, all crates in a compilation should use the same value of `-C patchable-function-entry` to reduce confusion.
```

Comment:
Is this the behavior we actually want? I realize many of our codegen flags work this way currently, but it seems undesirable for us to document this as the expectation for this feature. If I understand correctly, this lowers to a per-function attribute in LLVM so it should be possible for us to use the defining crate's `-Cpatchable-function-entry` value even when monomorphizing in a downstream crate.

Reply:
Monomorphization is pre-LLVM so we would need to include it in the rust metadata, but it seems doable in any case. I think the question is what's more desirable for the end user.

Reply:
It does lower to a per-function LLVM attribute, but we'd need to tag it onto all the embedded MIR or somewhere in the rmeta too. We could hack it in initially by implementing the flag as desugaring to an explicit attribute on all functions which don't already have one, so that it would be carried with the MIR into dependencies like an attribute would.

In the most common scenario, I expect `-C patchable-function-entry` will be set for the whole compilation, in which case either semantics works equivalently.

Aside from the surprise to compiler developers about a codegen flag being imported into a different codegen unit (within a certain context), there are two cases I can see where trying to follow the codegen flag for the defining crate would make a difference, for a pair of crates `patchable` and `unpatchable`, where `patchable` set the flag and `unpatchable` did not:

### `patchable` depends on `unpatchable`

Negatives:
- When building `patchable`, there will be some functions in the codegen'd `.o` file which are both not present in the `__patchable_function_entries` section of the binary

Positives:
- Allows you to set an entire crate as patchable or unpatchable, rather than needing to annotate individual functions

### `unpatchable` depends on `patchable`

Negatives:
- A crate that doesn't necessarily know about patchability ends up with a bonus section in its object file.
- If `unpatchable` is a `dylib`, `cdylib`, or `bin`, it owns a final link phase. This means that it may have stripping, objcopy, etc. which may not properly respect the presence of the `__patchable_function_entries` section. If `patchable` depended on that as an internal implementation detail, it won't work properly. Additionally, without  a linker script or similar shenanigans, `__patchable_function_entries` is unlikely to be mapped in a place usable by `patchable` if it needs it.

Positives:
- A crate could add hookability to its own functions, without the other side of the abstraction barrier needing to be aware (only works if functions are being referenced directly, as `__patchable_function_entries` won't be available.

## Conclusion

I think the main advantage of adding this would be the ability to flip patchability on or off at crate granularity. I'm not convinced this would ever be used, and if it did, it would be error prone.

I propose to punt on this and leave the flag effect as described above, or even loosen it further and state that behavior if `-C patchable-function-entry` is set to *different* values for different crates in a compilation graph, the result is either undefined or an error.

In the former case, we could add this feature in the future by making it possible to set this default via a crate-scoped or module-scoped attribute. In the latter, we leave any future behavior on the table, including what you suggested.

Reply:
> Aside from the surprise to compiler developers about a codegen flag being imported into a different codegen unit (within a certain context)

I agree that there could be some amount of surprise for compiler developers if the way this feature works diverges from other codegen properties but I don't think the way (for example) optimization flags or debuginfo levels are applied to code instantiated in downstream crates was ever specifically design that way, it's just a historical accident of implementation convenience. I believe that behavior runs contrary to what most users would expect and we should consider fixing it at some point. 

> there are two cases I can see where trying to follow the codegen flag for the defining crate would make a difference

These are good examples, thanks for coming up with them! The rationale you give for why these outcomes are confusing is interesting because it seems to me that there is an implicit connection being made between the presence of `-C patchable-function-entry` and whether the crate as a whole will have a `__patchable_function_entries` section consisting of all functions or not. Even in the model currently proposed by the RFC, this confusion can still exist as a crate compiled without `-C patchable-function-entry` can declare generic functions with `#[patchable(...)]` which will cause downstream crates to now contain patchable functions at the monomorphization instances. Likewise a generic `#[unpatchable]` function would do the reverse in downstream crates that compile with `-C patchable-function-entry`. 

I agree with you that the most likely way this will be used (and what we should recommend in documentation) is to compile all crates with the same settings which makes all of this a moot point. Punting on this specific detail seems appropriate and I would simply suggest removing the second sentence in this paragraph and allowing that to be an implementation detail (but very specifically *not* UB).